cni/pkg/nodeagent/informers.go

```diff
@@ -0,0 +1,259 @@
+			pods = append(pods, pod)
+		}
+	}
+	return pods
+}
+
+// EnqueueNamespace takes a Namespace and enqueues all Pod objects that make need an update
+// TODO it is sort of pointless/confusing/implicit to populate Old and New with the same reference here
+func (s *InformerHandlers) enqueueNamespace(o controllers.Object) {
+	namespace := o.GetName()
+	labels := o.GetLabels()
+	matchAmbient := labels[constants.DataplaneMode] == constants.DataplaneModeAmbient
+	if matchAmbient {
+		log.Infof("Namespace %s is enabled in ambient mesh", namespace)
+	} else {
+		log.Infof("Namespace %s is disabled from ambient mesh", namespace)
+	}
+	for _, pod := range s.pods.List(namespace, klabels.Everything()) {
+		// ztunnel pods are never "added to/removed from the mesh", so do not fire
+		// spurious events for them to avoid triggering extra
+		// ztunnel node reconciliation checks.
+		if !util.IsZtunnelPod(s.systemNamespace, pod) {
+			log.Debugf("Enqueuing pod %s/%s", pod.Namespace, pod.Name)
+			s.queue.Add(controllers.Event{
+				New:   pod,
+				Old:   pod,
+				Event: controllers.EventUpdate,
+			})
+		}
+	}
+}
+
+func (s *InformerHandlers) reconcile(input any) error {
+	event := input.(controllers.Event)
+	switch event.Latest().(type) {
+	case *corev1.Namespace:
+		return s.reconcileNamespace(input)
+	case *corev1.Pod:
+		return s.reconcilePod(input)
+	default:
+		return fmt.Errorf("unexpected event type: %+v", input)
+	}
+}
+
+func (s *InformerHandlers) reconcileNamespace(input any) error {
+	event := input.(controllers.Event)
+	ns := event.Latest().(*corev1.Namespace)
+
+	switch event.Event {
+	case controllers.EventAdd:
+		log.Debugf("Namespace %s added", ns.Name)
+		s.enqueueNamespace(ns)
+
+	case controllers.EventUpdate:
+		newNs := event.New.(*corev1.Namespace)
+		oldNs := event.Old.(*corev1.Namespace)
+
+		if getModeLabel(oldNs.Labels) != getModeLabel(newNs.Labels) {
+			log.Debugf("Namespace %s updated", newNs.Name)
+			s.enqueueNamespace(newNs)
+		}
+	}
+	return nil
+}
+
+func getModeLabel(m map[string]string) string {
+	if m == nil {
+		return ""
+	}
+	return m[constants.DataplaneMode]
+}
+
+func (s *InformerHandlers) reconcilePod(input any) error {
+	event := input.(controllers.Event)
+	pod := event.Latest().(*corev1.Pod)
+
+	defer EventTotals.With(eventTypeTag.Value(event.Event.String())).Increment()
+
+	switch event.Event {
+	case controllers.EventAdd:
+		// pod was added to our cache
+		// we get here in 2 cases:
+		// 1. new pod was created on our node
+		// 2. we were restarted and current existing pods are added to our cache
+
+		// We have no good way to distinguish between these two cases from here. But we don't need to!
+		// Existing pods will be handled by the dataplane using `GetAmbientPods`,
+		// and the initial enqueueNamespace, and new pods will be handled by the CNI.
+
+	case controllers.EventUpdate:
+		// For update, we just need to handle opt outs
+		newPod := event.New.(*corev1.Pod)
```

Comment:
should we use the latest pod in L177

Reply:
`Latest()` is just an alias for `event.New != nil return event.New else return event.Old` in `controllers/common.go`.


It's a little confusing as a helper method I agree, but at this point we don't know for sure that we have both an `old` and `new` event - later on, we do.